servlet/src/main/java/io/undertow/servlet/handlers/security/ServletFormAuthenticationMechanism.java

```diff
@@ -149,6 +155,37 @@ protected void handleRedirectBack(final HttpServerExchange exchange) {
 
     }
 
+    private static class FormResponseWrapper extends HttpServletResponseWrapper {
+
+        private int status = OK;
+
+        private FormResponseWrapper(HttpServletResponse response) {
+            super(response);
+        }
+
+        @Override
+        public void setStatus(int sc, String sm) {
+            if (super.getStatus() == OK) {
```

Comment:
I think this check should happen before applying the wrapper (i.e. if the current status code is 200 don't apply the wrapper, if it is not 200 don't let it be modified). Otherwise if there are multiple calls to setStatus the first one may work and the second will fail.